src/julia.jl

```diff
@@ -33,4 +33,56 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    first = f(a[1], a[1])
```

Comment:
The way to fix this is to check the type of `f(x,y)` in the loop, and if it doesn't match then re-allocate the matrix.   In the common case of type-stable functions `f`, the check will be eliminated by the compiler and it will be fast.

This is how comprehensions, `map`, and `broadcast` do it.